clean_user_registry_secrets/clean_user_registry_secrets.py

```diff
@@ -0,0 +1,151 @@
+        secret_name_match = re.match(secret_name_regex, secret_name)
+        # calculate secret age
+        tz = secret.metadata.creation_timestamp.tzinfo
+        secret_age = datetime.now(tz=tz) - secret.metadata.creation_timestamp
+        min_secret_age = timedelta(hours=min_secret_age_hrs)
+        if (
+            secret_name_match is not None
+            and secret.type == "kubernetes.io/dockerconfigjson"
+            and secret.metadata.labels.get("component") == "singleuser-server"
+            and reduce(  # check that label keys for sha, project and username are present
+                lambda acc, val: acc and (val in secret.metadata.labels.keys()),
+                label_keys,
+                True,
+            )
+        ):
+            podname = find_pod_by_secret(secret, k8s_client)
+            if podname is None:
+                # pod does not exist, delete if secret is old enough
+                if secret_age > min_secret_age:
+                    logging.info(
+                        f"User pod that used secret {secret_name} does not exist, "
+                        f"deleting secret as it is older "
+                        f"than the {min_secret_age_hrs} hours threshold"
+                    )
+                    k8s_client.delete_namespaced_secret(secret_name, namespace)
+            else:
+                # check if the pod has the expected annotations and is running or succeeded
+                # no need to check for secret age because we are certain secret has been used
+                pod = k8s_client.read_namespaced_pod(podname, namespace)
+                if (
+                    pod.metadata.labels.get("app") == "jupyterhub"
+                    and pod.metadata.labels.get("component")
+                    == "singleuser-server"
+                    and pod.status.phase in ["Running", "Succeeded"]
+                ):
+                    logging.info(
+                        f"Found user pod {podname} that used the secret, "
+                        f"deleting secret {secret_name}."
+                    )
+                    k8s_client.delete_namespaced_secret(secret_name, namespace)
+
+
+def float_gt_zero(number):
+    if float(number) <= 0:
+        raise argparse.ArgumentTypeError(
+            f"{number} should be a float and greater than zero."
+        )
+    else:
+        return float(number)
+
+
+def main():
+    # set logging level
+    logging.basicConfig(level=logging.INFO)
+
+    # check arguments
+    parser = argparse.ArgumentParser(description="Clean up user registry secrets.")
+    parser.add_argument(
+        "-n",
+        "--namespace",
+        type=str,
+        required=True,
+        help="K8s namespace where the user pods and registry secrets are located.",
+    )
+    parser.add_argument(
+        "-a",
+        "--age-hours-minimum",
+        type=float_gt_zero,
+        default=0.25,
+        help="The minimum age for a secret before it can be deleted if the user"
```

Comment:
see above - I find it confusing to refer to this as the minimum age when it's the max age a secret is allowed to have :)